util/compress/deflate.go

```diff
@@ -0,0 +1,174 @@
+	ErrOutOfMemory = errors.New("out of memory")
+	// ErrShortBuffer is returned when the output buffer is too small to fit the compressed/decompressed data
+	ErrShortBuffer = errors.New("short buffer")
+	// ErrNoInput is returned when no input buffer data is provided
+	ErrNoInput = errors.New("empty input")
+	// ErrBadData is returned when the compressed data is corrupted
+	ErrBadData = errors.New("data was corrupted, invalid or unsupported")
+	// ErrInsufficientSpace is returned when the provided output buffer is found to be too small during the decompression
+	ErrInsufficientSpace = errors.New("decompression failed: buffer too short. Retry with larger buffer")
+	// ErrShortOutput should not be generated using the current libdeflate usage, but remain here for compatibility
+	ErrShortOutput = errors.New("buffer too long: decompressed to fewer bytes than expected, indicating possible error in decompression. Make sure your out buffer has the exact length of the decompressed data or pass nil for out")
+	// ErrPartiallyConsumedData is returned if only a subset of the input data was consumed during the decompression
+	ErrPartiallyConsumedData = errors.New("partially consumed data")
+
+	// ErrUnknown returned when the libdeflate returns unexpected enum error
+	ErrUnknown = errors.New("unknown error code from decompressor library")
+)
+
+// Compress the input buffer into the output buffer.
+func Compress(in, out []byte, compressLevel int) (int, []byte, error) {
+	if len(in) == 0 {
+		return 0, out, ErrNoInput
+	}
+	if cap(out) == 0 {
+		return 0, out, ErrShortBuffer
+	}
+
+	if compressLevel < 1 {
+		compressLevel = 1
+	} else if compressLevel > 12 {
+		compressLevel = 12
+	}
+
+	c := C.libdeflate_alloc_compressor(C.int(compressLevel))
+	if C.isNull(unsafe.Pointer(c)) == 1 {
+		return 0, out, ErrOutOfMemory
+	}
+	defer func() {
+		C.libdeflate_free_compressor(c)
+	}()
+	inAddr := startMemAddr(in)
+	outAddr := startMemAddr(out)
+
+	written := int(C.libdeflate_gzip_compress(c, unsafe.Pointer(inAddr), C.size_t(len(in)), unsafe.Pointer(outAddr), C.size_t(cap(out))))
+
+	if written == 0 {
+		return written, out, ErrShortBuffer
+	}
+	return written, out[:written], nil
+}
+
+// Decompress decompresses the input buffer data into the output buffer.
+func Decompress(in, out []byte) ([]byte, error) {
+	if len(in) == 0 {
+		return out, ErrNoInput
+	}
+	if cap(out) == 0 {
+		return out, ErrShortBuffer
+	}
+	dc := C.libdeflate_alloc_decompressor()
+	if C.isNull(unsafe.Pointer(dc)) == 1 {
+		return out, ErrOutOfMemory
+	}
+	defer func() {
+		C.libdeflate_free_decompressor(dc)
+	}()
+
+	inAddr := startMemAddr(in)
+	outAddr := startMemAddr(out)
+
+	var actualInBytes C.size_t
+	var actualOutBytes C.size_t
+	r := C.libdeflate_gzip_decompress_ex(dc, unsafe.Pointer(inAddr), C.size_t(len(in)), unsafe.Pointer(outAddr), C.size_t(cap(out)), &actualInBytes, &actualOutBytes)
+
+	runtime.KeepAlive(&actualInBytes)
+	runtime.KeepAlive(&actualOutBytes)
```

Comment:
I'm not entirely sure that these are needed with the existing implementation, since the switch statements are using these variables, hence, making them accessible. Generally speaking, when passing local pointers to a C code, you need to add KeepAlive to ensure the garble collection won't clean these out while the C code is running.